### PR TITLE
Add Kubernetes support

### LINKED ISSUE #20: New module - DigitalOcean Kubernetes Cluster
##### SUMMARY

Add a new `digital_ocean_kubernetes_cluster` module.

##### ISSUE TYPE

- Feature Idea

##### COMPONENT NAME

N/A (new module)

##### ADDITIONAL INFORMATION

DigitalOcean allows management of managed Kubernetes clusters via the v2 API: https://developers.digitalocean.com/documentation/v2/#create-a-new-kubernetes-cluster

It would be nice if I could manage clusters in a DigitalOcean account using a native Ansible module. For example, to create a new cluster:

```yaml
- name: Ensure K8s cluster exists.
  digital_ocean_kubernetes_cluster:
    state: present
    name: mydroplet
    version: 1.12.1-do.2
    auto_upgrade: true
    region: sfo1
    node_pools:
      - size: s-1vcpu-2gb
        count: 3
        name: frontend-pool
      - size: c-4
        count: 2
        name: backend-pool
```

This issue was migrated from: https://github.com/ansible/ansible/issues/57244

### PR DESCRIPTION
**Pull request by mamercad**
##### SUMMARY

Add Kubernetes support (because everyone needs more Kubernetes 😄).

Fixes #20 

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

- digital_ocean_kubernetes
- digital_ocean_kubernetes_info

##### ADDITIONAL INFORMATION

**Creating a cluster:**

```bash
TASK [Create a new DigitalOcean Kubernetes cluster in New York 1] *****************************************************************************************************************************************************************************
task path: /Users/mmercado/src/github.com/mamercad/community.digitalocean/kubernetes.yml:8
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: mmercado
<localhost> EXEC /bin/sh -c 'echo ~mmercado && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/mmercado/.ansible/tmp `"&& mkdir /Users/mmercado/.ansible/tmp/ansible-tmp-1601574876.853937-38598-280848175700495 && echo ansible-tmp-1601574876.853937-38598-280848175700495="` echo /Users/mmercado/.ansible/tmp/ansible-tmp-1601574876.853937-38598-280848175700495 `" ) && sleep 0'
Using module file /Users/mmercado/.ansible/collections/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_kubernetes.py
<localhost> PUT /Users/mmercado/.ansible/tmp/ansible-local-38555w6hj4zxo/tmp7henyl7d TO /Users/mmercado/.ansible/tmp/ansible-tmp-1601574876.853937-38598-280848175700495/AnsiballZ_digital_ocean_kubernetes.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/mmercado/.ansible/tmp/ansible-tmp-1601574876.853937-38598-280848175700495/ /Users/mmercado/.ansible/tmp/ansible-tmp-1601574876.853937-38598-280848175700495/AnsiballZ_digital_ocean_kubernetes.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /Users/mmercado/.ansible/tmp/ansible-tmp-1601574876.853937-38598-280848175700495/AnsiballZ_digital_ocean_kubernetes.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /Users/mmercado/.ansible/tmp/ansible-tmp-1601574876.853937-38598-280848175700495/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true 
  data:
    kubeconfig: |-
      apiVersion: v1
      clusters:
      - cluster:
          certificate-authority-data: REDACTED
          server: https://REDACTED.k8s.ondigitalocean.com
        name: do-nyc1-hacktoberfest
      contexts:
      - context:
          cluster: do-nyc1-hacktoberfest
          user: do-nyc1-hacktoberfest-admin
        name: do-nyc1-hacktoberfest
      current-context: do-nyc1-hacktoberfest
      kind: Config
      preferences: {}
      users:
      - name: do-nyc1-hacktoberfest-admin
        user:
          token: REDACTED
    kubernetes_cluster:
      auto_upgrade: false
      cluster_subnet: 10.244.0.0/16
      created_at: '2020-10-01T17:54:41Z'
      endpoint: https://REDACTED.k8s.ondigitalocean.com
      id: REDACTED
      ipv4: REDACTED
      maintenance_policy:
        day: any
        duration: 4h0m0s
        start_time: '7:00'
      name: hacktoberfest
      node_pools:
      - auto_scale: false
        count: 3
        id: d594e929-d434-4a32-87e3-6d5f0d7e4309
        labels: null
        max_nodes: 0
        min_nodes: 0
        name: hacktoberfest-workers
        nodes:
        - created_at: '2020-10-01T17:54:41Z'
          droplet_id: '210186393'
          id: c6c24d1f-6445-4a06-bcaf-b7edf0794f40
          name: hacktoberfest-workers-3l6yx
          status:
            state: running
          updated_at: '2020-10-01T17:58:43Z'
        - created_at: '2020-10-01T17:54:41Z'
          droplet_id: '210186395'
          id: 9a75659e-a1bc-4e68-b2c7-13e299c53410
          name: hacktoberfest-workers-3l6yo
          status:
            state: running
          updated_at: '2020-10-01T17:59:21Z'
        - created_at: '2020-10-01T17:54:41Z'
          droplet_id: '210186394'
          id: 4dc4c54f-8bda-4787-bc6a-4bf28c0e073f
          name: hacktoberfest-workers-3l6yj
          status:
            state: running
          updated_at: '2020-10-01T17:59:21Z'
        size: s-1vcpu-2gb
        tags:
        - k8s
        - k8s:REDACTED
        - k8s:worker
        taints: []
      region: nyc1
      service_subnet: 10.245.0.0/16
      status:
        state: running
      surge_upgrade: false
      tags:
      - k8s
      - k8s:REDACTED
      updated_at: '2020-10-01T18:02:03Z'
      version: 1.18.8-do.1
      vpc_uuid: 0c20a2e6-b9ca-43fd-8b22-def8741321ac
  invocation:
    module_args:
      auto_upgrade: false
      maintenance_policy: null
      name: hacktoberfest
      node_pools:
      - count: 3
        name: hacktoberfest-workers
        size: s-1vcpu-2gb
      region: nyc1
      surge_upgrade: false
      tags: null
      version: 1.18.8-do.1
      vpc_uuid: null

TASK [Show the kubeconfig for the cluster we just created] ************************************************************************************************************************************************************************************
task path: /Users/mmercado/src/github.com/mamercad/community.digitalocean/kubernetes.yml:22
ok: [localhost] => 
  msg: |-
    apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: REDACTED
        server: https://REDACTED.k8s.ondigitalocean.com
      name: do-nyc1-hacktoberfest
    contexts:
    - context:
        cluster: do-nyc1-hacktoberfest
        user: do-nyc1-hacktoberfest-admin
      name: do-nyc1-hacktoberfest
    current-context: do-nyc1-hacktoberfest
    kind: Config
    preferences: {}
    users:
    - name: do-nyc1-hacktoberfest-admin
      user:
        token: REDACTED
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Inspecting a cluster:**

```bash
TASK [Get information about an existing DigitalOcean Kubernetes cluster] **********************************************************************************************************************************************************************
task path: /Users/mmercado/src/github.com/mamercad/community.digitalocean/kubernetes.yml:32
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: mmercado
<localhost> EXEC /bin/sh -c 'echo ~mmercado && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/mmercado/.ansible/tmp `"&& mkdir /Users/mmercado/.ansible/tmp/ansible-tmp-1601575631.404768-69736-169422990869984 && echo ansible-tmp-1601575631.404768-69736-169422990869984="` echo /Users/mmercado/.ansible/tmp/ansible-tmp-1601575631.404768-69736-169422990869984 `" ) && sleep 0'
Using module file /Users/mmercado/.ansible/collections/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_kubernetes_info.py
<localhost> PUT /Users/mmercado/.ansible/tmp/ansible-local-69693bqn2beos/tmpnkvpl1rz TO /Users/mmercado/.ansible/tmp/ansible-tmp-1601575631.404768-69736-169422990869984/AnsiballZ_digital_ocean_kubernetes_info.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/mmercado/.ansible/tmp/ansible-tmp-1601575631.404768-69736-169422990869984/ /Users/mmercado/.ansible/tmp/ansible-tmp-1601575631.404768-69736-169422990869984/AnsiballZ_digital_ocean_kubernetes_info.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /Users/mmercado/.ansible/tmp/ansible-tmp-1601575631.404768-69736-169422990869984/AnsiballZ_digital_ocean_kubernetes_info.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /Users/mmercado/.ansible/tmp/ansible-tmp-1601575631.404768-69736-169422990869984/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true 
  data:
    auto_upgrade: false
    cluster_subnet: 10.244.0.0/16
    created_at: '2020-10-01T17:54:41Z'
    endpoint: https://REDACTED.k8s.ondigitalocean.com
    id: REDACTED
    ipv4: 159.65.239.183
    kubeconfig: |-
      apiVersion: v1
      clusters:
      - cluster:
          certificate-authority-data: REDACTED
          server: https://REDACTED.k8s.ondigitalocean.com
        name: do-nyc1-hacktoberfest
      contexts:
      - context:
          cluster: do-nyc1-hacktoberfest
          user: do-nyc1-hacktoberfest-admin
        name: do-nyc1-hacktoberfest
      current-context: do-nyc1-hacktoberfest
      kind: Config
      preferences: {}
      users:
      - name: do-nyc1-hacktoberfest-admin
        user:
          token: REDACTED
    maintenance_policy:
      day: any
      duration: 4h0m0s
      start_time: '7:00'
    name: hacktoberfest
    node_pools:
    - auto_scale: false
      count: 3
      id: d594e929-d434-4a32-87e3-6d5f0d7e4309
      labels: null
      max_nodes: 0
      min_nodes: 0
      name: hacktoberfest-workers
      nodes:
      - created_at: '2020-10-01T17:54:41Z'
        droplet_id: '210186393'
        id: c6c24d1f-6445-4a06-bcaf-b7edf0794f40
        name: hacktoberfest-workers-3l6yx
        status:
          state: running
        updated_at: '2020-10-01T17:58:43Z'
      - created_at: '2020-10-01T17:54:41Z'
        droplet_id: '210186395'
        id: 9a75659e-a1bc-4e68-b2c7-13e299c53410
        name: hacktoberfest-workers-3l6yo
        status:
          state: running
        updated_at: '2020-10-01T17:59:21Z'
      - created_at: '2020-10-01T17:54:41Z'
        droplet_id: '210186394'
        id: 4dc4c54f-8bda-4787-bc6a-4bf28c0e073f
        name: hacktoberfest-workers-3l6yj
        status:
          state: running
        updated_at: '2020-10-01T17:59:21Z'
      size: s-1vcpu-2gb
      tags:
      - k8s
      - k8s:REDACTED
      - k8s:worker
      taints: []
    region: nyc1
    service_subnet: 10.245.0.0/16
    status:
      state: running
    surge_upgrade: false
    tags:
    - k8s
    - k8s:REDACTED
    updated_at: '2020-10-01T18:02:03Z'
    version: 1.18.8-do.1
    vpc_uuid: 0c20a2e6-b9ca-43fd-8b22-def8741321ac
  invocation:
    module_args:
      name: hacktoberfest

TASK [debug] **********************************************************************************************************************************************************************************************************************************
task path: /Users/mmercado/src/github.com/mamercad/community.digitalocean/kubernetes.yml:39
ok: [localhost] => 
  msg: Cluster name is hacktoberfest, ID is REDACTED

TASK [debug] **********************************************************************************************************************************************************************************************************************************
task path: /Users/mmercado/src/github.com/mamercad/community.digitalocean/kubernetes.yml:42
ok: [localhost] => 
  msg: |-
    Cluster kubeconfig is apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: REDACTED
        server: https://REDACTED.k8s.ondigitalocean.com
      name: do-nyc1-hacktoberfest
    contexts:
    - context:
        cluster: do-nyc1-hacktoberfest
        user: do-nyc1-hacktoberfest-admin
      name: do-nyc1-hacktoberfest
    current-context: do-nyc1-hacktoberfest
    kind: Config
    preferences: {}
    users:
    - name: do-nyc1-hacktoberfest-admin
      user:
        token: REDACTED
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Destroying a cluster:**

```bash
TASK [Destroy (delete) an existing DigitalOcean Kubernetes cluster] ***************************************************************************************************************************************************************************
task path: /Users/mmercado/src/github.com/mamercad/community.digitalocean/kubernetes.yml:26
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: mmercado
<localhost> EXEC /bin/sh -c 'echo ~mmercado && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/mmercado/.ansible/tmp `"&& mkdir /Users/mmercado/.ansible/tmp/ansible-tmp-1601575750.572343-73653-30787478886142 && echo ansible-tmp-1601575750.572343-73653-30787478886142="` echo /Users/mmercado/.ansible/tmp/ansible-tmp-1601575750.572343-73653-30787478886142 `" ) && sleep 0'
Using module file /Users/mmercado/.ansible/collections/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_kubernetes.py
<localhost> PUT /Users/mmercado/.ansible/tmp/ansible-local-73610xopa3feu/tmpmsam_491 TO /Users/mmercado/.ansible/tmp/ansible-tmp-1601575750.572343-73653-30787478886142/AnsiballZ_digital_ocean_kubernetes.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/mmercado/.ansible/tmp/ansible-tmp-1601575750.572343-73653-30787478886142/ /Users/mmercado/.ansible/tmp/ansible-tmp-1601575750.572343-73653-30787478886142/AnsiballZ_digital_ocean_kubernetes.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /Users/mmercado/.ansible/tmp/ansible-tmp-1601575750.572343-73653-30787478886142/AnsiballZ_digital_ocean_kubernetes.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /Users/mmercado/.ansible/tmp/ansible-tmp-1601575750.572343-73653-30787478886142/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true 
  invocation:
    module_args:
      auto_upgrade: false
      maintenance_policy: null
      name: hacktoberfest
      node_pools:
      - count: 1
        labels: {}
        name: worker-pool
        size: s-1vcpu-2gb
        tags: []
      region: nyc1
      surge_upgrade: false
      tags: null
      version: 1.18.8-do.1
      vpc_uuid: null
  msg: Kubernetes cluster deleted
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### CODE REVIEW

**Comment by geerlingguy** [comment]
Just wanted to note that I'm very interested in getting this merged. And I'm glad you're living your best docstring life :D

**Comment by mamercad** [comment]
> Just wanted to note that I'm very interested in getting this merged. And I'm glad you're living your best docstring life :D

Heh, it's actually driving me crazy a bit, but, that's the dev/ops life :)

File: plugins/modules/digital_ocean_kubernetes.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    type: int
+    default: 600
+requirements:
+    - python >= 2.6
+'''
+
+
+EXAMPLES = r'''
+- name: Create a new DigitalOcean Kubernetes cluster in New York 1
+  community.digitalocean.digital_ocean_kubernetes:
```
Comment (geerlingguy) [inline plugins/modules/digital_ocean_kubernetes.py:114]: Just a quick thought: Since this is the first module being added to the collection independent of Ansible core... should we name the module `kubernetes` so it's FQCN is `community.digitalocean.kubernetes`?

File: plugins/modules/digital_ocean_kubernetes.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+- name: Destroy (delete) an existing DigitalOcean Kubernetes cluster
+  community.digitalocean.digital_ocean_kubernetes:
+    state: absent
+    oauth_token: "{{ lookup('env', 'DO_API_TOKEN') }}"
+    name: hacktoberfest
+'''
+
+
+# Digital Ocean API info https://developers.digitalocean.com/documentation/v2/#kubernetes
+# The only variance from the documented response is that the kubeconfig is (if return_kubeconfig is True) merged in at data['kubeconfig']
```
Comment (geerlingguy) [inline plugins/modules/digital_ocean_kubernetes.py:140]: I'm wondering if we really want to return the kubeconfig, or if it's better practice to get the kubeconfig in a separate explicit task? Genuinely wondering, because I know with EKS in AWS, I generally have a separate task after the cluster is created to generate/retrieve the kubeconfig using the AWS CLI (since there isn't a direct Ansible module for it).

File: plugins/modules/digital_ocean_kubernetes.py (hunk 1 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_kubernetes.py:114]: 👍

File: plugins/modules/digital_ocean_kubernetes.py (hunk 2 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_kubernetes.py:140]: ```            return_kubeconfig=dict(type='bool', default=False),```

Was my thinking ... I'm fine either way though. Few more passes through the docstring battle and I'm going to 💤 .

File: plugins/modules/digital_ocean_kubernetes.py (hunk 1 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_kubernetes.py:114]: I'll hack this change out soon.

**Comment by mamercad** [comment]
Hrm, wonder why I'm getting different results locally ...

```
Running sanity test 'validate-modules' with Python 3.6
Run command: /usr/bin/python3.6 /root/ansible/test/lib/ansible_test/_data/sanity/validate-modules/validate-modules --format json --arg-spec plugins/modules/digital_ocean.py plugins/modules/digital_ocean_account_facts.py plugins/modules/digital_oc ...
ERROR: Found 12 validate-modules issue(s) which need to be resolved:
ERROR: plugins/modules/digital_ocean_account_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_account_info'
ERROR: plugins/modules/digital_ocean_certificate_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_certificate_info'
ERROR: plugins/modules/digital_ocean_domain_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_domain_info'
ERROR: plugins/modules/digital_ocean_firewall_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_firewall_info'
ERROR: plugins/modules/digital_ocean_floating_ip_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_floating_ip_info'
ERROR: plugins/modules/digital_ocean_image_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_image_info'
ERROR: plugins/modules/digital_ocean_load_balancer_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_load_balancer_info'
ERROR: plugins/modules/digital_ocean_region_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_region_info'
ERROR: plugins/modules/digital_ocean_size_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_size_info'
ERROR: plugins/modules/digital_ocean_snapshot_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_snapshot_info'
ERROR: plugins/modules/digital_ocean_tag_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_tag_info'
ERROR: plugins/modules/digital_ocean_volume_facts.py:0:0: invalid-documentation: DOCUMENTATION.module: not a valid value for dictionary value @ data['module']. Got 'digital_ocean_volume_info'
```

**Comment by mamercad** [comment]
Alright, I don't know where to go from here, I'm in some sort of Catch-22 that I don't know how to get out of:

```
❯ grep state=dict plugins/modules/digital_ocean_kubernetes.py
            state=dict(choices=['present', 'absent'], default='present'),
❯ ansible-test sanity --test validate-modules 2>&1 | grep kuber | grep state
ERROR: plugins/modules/digital_ocean_kubernetes.py:0:0: doc-required-mismatch: Argument 'state' in argument_spec is not required, but is documented as being required
```

```
❯ grep state=dict plugins/modules/digital_ocean_kubernetes.py
            state=dict(choices=['present', 'absent'], default='present', required=True),
❯ ansible-test sanity --test validate-modules 2>&1 | grep kuber | grep state
ERROR: plugins/modules/digital_ocean_kubernetes.py:0:0: invalid-ansiblemodule-schema: AnsibleModule.argument_spec.state: required=True cannot be supplied with a default for dictionary value @ data['argument_spec']['state']. Got {'choices': ['present', 'absent'], 'default': 'present', 'required': True}
ERROR: plugins/modules/digital_ocean_kubernetes.py:0:0: no-default-for-required-parameter: Argument 'state' in argument_spec is marked as required but specifies a default. Arguments with a default should not be marked as required
```

File: plugins/modules/digital_ocean_kubernetes.py (hunk 1 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_kubernetes.py]: https://github.com/ansible-collections/community.digitalocean/pull/25/commits/c9d76ddefed0406ca3eb1ba5c0ddc255ecac7dfd

**Comment by mamercad** [comment]
Finally got the builds to pass. Going to make sure it still *works*.

**Comment by mamercad** [comment]
```
❯ ansible-playbook -i localhost, -c local kubernetes.yml
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features under development. This is a rapidly changing source of code and can
become unstable at any point.
[WARNING]: running playbook inside collection community.digitalocean

PLAY [kubernetes] *******************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************************************************************************
[WARNING]: Platform darwin on host localhost is using the discovered Python interpreter at /usr/bin/python, but future installation of another Python interpreter could change the meaning of that path. See
https://docs.ansible.com/ansible/devel/reference_appendices/interpreter_discovery.html for more information.
ok: [localhost]

TASK [Create a new DigitalOcean Kubernetes cluster in New York 1] *******************************************************************************************************************************************************************************
changed: [localhost]

TASK [Get information about an existing DigitalOcean Kubernetes cluster] ************************************************************************************************************************************************************************
changed: [localhost]

TASK [debug] ************************************************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg: Cluster name is hacktoberfest, ID is 0dda1738-e647-4283-aab7-76ea2a39cd1f

TASK [debug] ************************************************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg: |-
    Cluster kubeconfig is apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: REDACTED
        server: https://0dda1738-e647-4283-aab7-76ea2a39cd1f.k8s.ondigitalocean.com
      name: do-nyc1-hacktoberfest
    contexts:
    - context:
        cluster: do-nyc1-hacktoberfest
        user: do-nyc1-hacktoberfest-admin
      name: do-nyc1-hacktoberfest
    current-context: do-nyc1-hacktoberfest
    kind: Config
    preferences: {}
    users:
    - name: do-nyc1-hacktoberfest-admin
      user:
        token: REDACTED

TASK [pause] ************************************************************************************************************************************************************************************************************************************
Pausing for 60 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [localhost]

TASK [Destroy (delete) an existing DigitalOcean Kubernetes cluster] *****************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP **************************************************************************************************************************************************************************************************************************************
localhost                  : ok=7    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

File: plugins/modules/digital_ocean_kubernetes.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+      - The slug identifier for the version of Kubernetes used for the cluster. See the /v2/kubernetes/options endpoint for available versions.
+    type: str
+    required: no
+    default: 1.18.8-do.1
+  auto_upgrade:
+    description:
+      - A boolean value indicating whether the cluster will be automatically upgraded to new patch releases during its maintenance window.
+    type: bool
+    required: no
+    default: False
```
Comment (felixfontein) [inline plugins/modules/digital_ocean_kubernetes.py:53]:
```suggestion
    default: false
```
The general recommendation is to only use `true` and `false` in documentation.

File: plugins/modules/digital_ocean_kubernetes.py [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+    description:
+      - The slug identifier for the version of Kubernetes used for the cluster. See the /v2/kubernetes/options endpoint for available versions.
+    type: str
+    required: no
+    default: 1.18.8-do.1
+  auto_upgrade:
+    description:
+      - A boolean value indicating whether the cluster will be automatically upgraded to new patch releases during its maintenance window.
+    type: bool
+    required: no
```
Comment (felixfontein) [inline plugins/modules/digital_ocean_kubernetes.py:52]:
```suggestion
```
Another recommendation is to don't specify default values. Feel free to ignore this one if you prefer to always have them there. But if you want them to be there, better use:
```suggestion
    required: false
```

File: plugins/modules/digital_ocean_kubernetes_info.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    type: str
+    required: yes
+  return_kubeconfig:
+    description:
+      - Controls whether or not to return the C(kubeconfig).
+    type: bool
+    required: no
+    default: False
+requirements:
+  - python >= 2.6
```
Comment (felixfontein) [inline plugins/modules/digital_ocean_kubernetes_info.py:37]:
```suggestion
```
Ansible 2.9 and later do not support Python before 2.6 anyway, so it's not really necessary to mention that.

File: plugins/modules/digital_ocean_kubernetes_info.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+EXAMPLES = r'''
+- name: Get information about an existing DigitalOcean Kubernetes cluster
+  community.digitalocean.digital_ocean_kubernetes_info:
+    oauth_token: "{{ lookup('env', 'DO_API_TOKEN') }}"
+    name: hacktoberfest
+    return_kubeconfig: yes
+  register: my_cluster
+
+- debug:
```
Comment (felixfontein) [inline plugins/modules/digital_ocean_kubernetes_info.py:49]:
```suggestion
- ansible.builtin.debug:
```

File: plugins/modules/digital_ocean_kubernetes_info.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    default: False
+requirements:
+  - python >= 2.6
+'''
+
+
+EXAMPLES = r'''
+- name: Get information about an existing DigitalOcean Kubernetes cluster
+  community.digitalocean.digital_ocean_kubernetes_info:
+    oauth_token: "{{ lookup('env', 'DO_API_TOKEN') }}"
```
Comment (felixfontein) [inline plugins/modules/digital_ocean_kubernetes_info.py:44]:
```suggestion
    oauth_token: "{{ lookup('ansible.builtin.env', 'DO_API_TOKEN') }}"
```

**Review by felixfontein — COMMENTED** [review]
Just some general remarks from glancing over this.

File: plugins/modules/digital_ocean_kubernetes_info.py [hunk 8]
```diff
@@ -21,35 +21,33 @@
...
+    required: true
   return_kubeconfig:
     description:
       - Controls whether or not to return the C(kubeconfig).
     type: bool
-    required: no
-    default: False
-requirements:
-  - python >= 2.6
+    required: true
```
Comment (felixfontein) [inline plugins/modules/digital_ocean_kubernetes_info.py]: This wasn't required before. Is this change intentional?

Comment (mamercad) [reply plugins/modules/digital_ocean_kubernetes_info.py]: Nope, not required, currently fixed (defaulted to `false`).

File: plugins/modules/digital_ocean_kubernetes.py (hunk 1 above)
Comment (felixfontein) [reply plugins/modules/digital_ocean_kubernetes.py:130]: What did happen to this?

Comment (mamercad) [reply plugins/modules/digital_ocean_kubernetes.py:130]: After not looking at this code for a while, and looking again the other day, I had a change of heart wrt dropping the "redundant digital_ocean prefix", so, I'm making this module match everything else in this collection. Let's give this some thought for 2.0.0, i.e., should things look like `community.digitalocean.digital_ocean_kubernetes` or more simply `community.digitalocean.kubernetes` (applying this same format to every module in here). Personally, I like the latter.

**Comment by mamercad** [comment]
Rolled into #75.
